src/com/facebook/buck/jvm/kotlin/KotlinBuckConfig.java

```diff
@@ -58,67 +63,105 @@ public KotlinBuckConfig(BuckConfig delegate) {
 
-    if (value.isPresent()) {
-      boolean isAbsolute = Paths.get(value.get()).isAbsolute();
-      if (isAbsolute) {
-        return Either.ofRight(delegate.getPath(SECTION, "runtime_jar", false).get().normalize());
-      } else {
-        return Either.ofLeft(delegate.getSourcePath(SECTION, "runtime_jar").get());
-      }
+    stdlib = getKotlinHome().resolve(Paths.get("lib", "kotlin-runtime.jar"));
+    if (Files.isRegularFile(stdlib)) {
+      return stdlib.normalize();
+    }
+
+    stdlib = getKotlinHome().resolve(Paths.get("libexec","lib", "kotlin-runtime.jar"));
+    if (Files.isRegularFile(stdlib)) {
+      return stdlib.normalize();
+    }
+
+    throw new HumanReadableException("Could not resolve kotlin stdlib JAR location (kotlin home:"
+        + getKotlinHome() + ").");
+  }
+
+  /**
+   * Get the path to the Kotlin compiler jar.
+   * @return the Kotlin compiler jar path
+   */
+  Path getPathToCompilerJar() {
+    Path compiler = getKotlinHome().resolve("kotlin-compiler.jar");
+    if (Files.isRegularFile(compiler)) {
+      return compiler.normalize();
     }
 
-    Path runtime = getKotlinHome().resolve("kotlin-runtime.jar");
-    if (Files.isRegularFile(runtime)) {
-      return Either.ofRight(runtime.toAbsolutePath().normalize());
+    compiler = getKotlinHome().resolve(Paths.get("lib", "kotlin-compiler.jar"));
+    if (Files.isRegularFile(compiler)) {
+      return compiler.normalize();
     }
 
-    runtime = getKotlinHome().resolve(Paths.get("lib", "kotlin-runtime.jar"));
-    if (Files.isRegularFile(runtime)) {
-      return Either.ofRight(runtime.toAbsolutePath().normalize());
+    compiler = getKotlinHome().resolve(Paths.get("libexec","lib", "kotlin-compiler.jar"));
+    if (Files.isRegularFile(compiler)) {
+      return compiler.normalize();
     }
 
-    throw new HumanReadableException("Could not resolve kotlin runtime JAR location.");
+    throw new HumanReadableException("Could not resolve kotlin compiler JAR location (kotlin home:"
+        + getKotlinHome() + ").");
+  }
+
+  /**
+   * Determine whether external Kotlin compilation is being forced.  The default is
+   * internal (in-process) execution, but this can be overridden in .buckconfig by
+   * setting the "external" property to "true".
+   *
+   * @return true is external compilation is requested, false otherwise
+   */
+  private boolean isExternalCompilation() {
+    Optional<Boolean> value = delegate.getBoolean(SECTION, "external");
+    return value.orElse(false);
   }
 
+
+  /**
+   * Find the Kotlin home (installation) directory by searching in this order: <br>
+   *  <ul>
+   *    <li>If the "kotlin_home" directory is specified in .buckconfig then use it.</li>
+   *    <li>Check the environment for a KOTLIN_HOME variable, if defined then use it.</li>
+   *    <li>Resolve "kotlinc" with an ExecutableFinder, and if found then deduce the
+   *        kotlin home directory from it.</li>
+   *  </ul>
+   *
+   * @return the Kotlin home path
+   */
```

Comment:
The types of directory structures we consider valid kotlin_home isn't quite clear.  If I have my kotlinc in `~/bin/kotlinc` then this considers `~/bin` to be my kotlin_home which seems like it'd be wrong (`~/bin/lib/kotlin-runtime.jar`? no thanks).

Can we add a verify method that makes sure `kotlin_home`s are either structured like:
```
kotlin_home_as_downloaded_and_unziped_from_website/
    bin/
    lib/
```
or 
```
kotlin_home_as_installed_from_homebrew/
    bin/
    libexec/
        lib/
```